google/cloud/storage/internal/service_account_credentials.h

```diff
@@ -0,0 +1,183 @@
+    // potential drawbacks to this approach.
+    std::string token_uri;
+    const char TOKEN_URI_KEY[] = "token_uri";
+    if (credentials.count(TOKEN_URI_KEY)) {
+      token_uri = credentials["token_uri"].get_ref<std::string const&>();
+    } else {
+      token_uri = GoogleOAuthRefreshEndpoint();
+    }
+    long int cur_time = time_fetcher_.GetSecondsSinceEpoch();
+    nl::json assertion_payload = {
+        {"iss", credentials["private_key_id"].get_ref<std::string const&>()},
+        {"scope", scope_oss.str()},
+        {"aud", token_uri},
+        {"iat", cur_time},
+        // Resulting access token should be expire after one hour.
+        {"exp", cur_time + GoogleOAuthAccessTokenLifetime()}};
+
+    std::string svc_acct_private_key_pem =
+        credentials["private_key"].get_ref<std::string const&>();
+    std::string payload(
+        requestor_
+            .MakeEscapedString(
+                "grant_type=urn:ietf:params:oauth:grant-type:jwt-bearer")
+            .get());
+    payload += "&assertion=";
+    payload += MakeJWTAssertion(assertion_header, assertion_payload,
+                                svc_acct_private_key_pem);
+    requestor_.PrepareRequest(std::move(payload));
+  }
+
+  std::string AuthorizationHeader() override {
+    std::unique_lock<std::mutex> lk(mu_);
+    cv_.wait(lk, [this]() { return Refresh(); });
+    return authorization_header_;
+  }
+
+ private:
+  std::string MakeJWTAssertion(const nl::json& header, const nl::json& payload,
+                               const std::string& pem_contents) {
+    std::string encoded_header =
+        OpenSslUtils::UrlsafeBase64Encode(header.dump());
+    std::string encoded_payload =
+        OpenSslUtils::UrlsafeBase64Encode(payload.dump());
+    std::string encoded_signature =
+        OpenSslUtils::UrlsafeBase64Encode(OpenSslUtils::SignStringWithPem(
+            encoded_header + '.' + encoded_payload, pem_contents,
+            JWTSigningAlgorithms::RS256));
+    return encoded_header + '.' + encoded_payload + '.' + encoded_signature;
+  }
+
+  bool Refresh() {
+    if (std::chrono::system_clock::now() < expiration_time_) {
+      return true;
+    }
+
+    // TODO(#516) - use retry policies to refresh the credentials.
+    auto response = requestor_.MakeRequest();
+    if (200 != response.status_code) {
+      return false;
+    }
+
+    nl::json access_token = nl::json::parse(response.payload);
+    // Response should have the attributes "access_token", "expires_in", and
+    // "token_type".
+    std::ostringstream header_oss;
+    header_oss << "Authorization: "
+               << access_token["token_type"].get_ref<std::string const&>()
+               << ' '
+               << access_token["access_token"].get_ref<std::string const&>();
+    std::string header = header_oss.str();
+    auto expires_in = std::chrono::seconds(access_token["expires_in"]);
+
+    auto slack_multiplier = RefreshTimeSlackPercent();
```

Comment:
Please follow the canonical implementation and use a buffer period of a constant 5 minutes: https://github.com/GoogleCloudPlatform/google-auth-library-python/blob/master/google/auth/_helpers.py#L25

Reply:
Fixed, here and in the authorized_user_credentials header.